### PR TITLE
implemented independentTreeShap algorithm

### PR DESCRIPTION
**Pull request by IvanKozlov98**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=ru.

### CODE REVIEW

File: catboost/libs/fstr/shap_values.cpp [hunk 1]
```diff
@@ -721,7 +867,176 @@ static void MapBinFeaturesToClasses(
...
+                independentTreeShapParams.ReferenceIndicesForAllTrees[treeIdx],
+                independentTreeShapParams.ReferenceLeafIndicesForAllTrees[treeIdx],
+                treeIdx,
+                /*isCalcForAllLeaves*/ false,
+                &shapValuesForAllReferences,
+                &meanValueForAllReferences
+            );
+        } else {
+            const size_t leafCount = (size_t(1) << forest.GetTreeSizes()[treeIdx]);
+            TVector<TVector<TVector<double>>> shapValueByDepthBetweenLeafes(leafCount);
```
Comment (Evgueni-Petrov-aka-espetrov) [inline catboost/libs/fstr/shap_values.cpp]: shapValueByDepthBetweenLeafes --> shapValueByDepthBetweenLeaves

File: catboost/libs/fstr/shap_values.cpp [hunk 2]
```diff
@@ -141,6 +143,117 @@ static void UpdateShapByFeaturePath(
...
+
+ // Independent Tree SHAP
+TContribution TInternalIndependentTreeShapCalcer::Calc(
+    int depth,
+    size_t nodeIdx,
+    ui32 uniqueFeaturesCount,
+    ui32 featureMatchedDatasetSampleCount
+) {
+    const auto approxDimension = Forest.GetDimensionsCount();
+    const double* leafValuesPtr = Forest.GetFirstLeafPtrForTree(TreeIdx);
```
Comment (Evgueni-Petrov-aka-espetrov) [inline catboost/libs/fstr/shap_values.cpp]: for faster performance, call Forest.GetDimensionsCount(), Forest.GetFirstLeafPtrForTree(TreeIdx), Forest.GetTreeSizes()[TreeIdx] once in calcer's constructor, and store the results in const members, and use these stored results in Calc.

use const array ref to store result of Forest.GetTreeSplits()

File: catboost/libs/fstr/shap_values.cpp [hunk 3]
```diff
@@ -141,6 +143,117 @@ static void UpdateShapByFeaturePath(
...
+                        * leafValuesPtr[nodeIdx * approxDimension + dimension];
+                }
+            }
+        }
+        return contribution;
+    }
+    const size_t remainingDepth = Forest.GetTreeSizes()[TreeIdx] - depth - 1;
+    const int combinationClass = BinFeatureCombinationClass[
+        Forest.GetTreeSplits()[Forest.GetTreeStartOffsets()[TreeIdx] + remainingDepth]
+    ];
```
Comment (Evgueni-Petrov-aka-espetrov) [inline catboost/libs/fstr/shap_values.cpp]: for faster performance, fill BinFeatureCombinationClass for all possible depths in constructor

File: catboost/libs/fstr/shap_values.cpp [hunk 4]
```diff
@@ -141,6 +143,117 @@ static void UpdateShapByFeaturePath(
...
+    constexpr size_t None = (size_t)(-1);
+    size_t nextNodeIdx = None;
+    // for current object, who fall down to leadIdx [documentLeafIdx]
+    const bool isGoRight = (DocumentLeafIdx >> remainingDepth) & 1;
+    const size_t goNodeIdx = nodeIdx * 2 + isGoRight;
+    // for current reference object, who fall down to leadIdx [documentLeafIdx]
+    const bool isGoRightRefernce = (DocumentLeafIdxReference >> remainingDepth) & 1;
+    const size_t goNodeIdxRefernce = nodeIdx * 2 + isGoRightRefernce;
```
Comment (Evgueni-Petrov-aka-espetrov) [inline catboost/libs/fstr/shap_values.cpp]: isGoRightRefernce --> isGoRightReference
goNodeIdxRefernce --> goNodeIdxReference

File: catboost/libs/fstr/shap_values.cpp [hunk 5]
```diff
@@ -6,9 +6,11 @@
 #include <catboost/private/libs/algo/index_calcer.h>
 #include <catboost/libs/data/features_layout.h>
 #include <catboost/libs/helpers/exception.h>
+#include <catboost/libs/helpers/mem_usage.h>
```
[the remark this answers is not in the document]
Comment (IvanKozlov98) [reply catboost/libs/fstr/shap_values.cpp]: use function GetMonopolisticFreeCpuRam from mem_usage.h

**Pull request closed without merging**